File: red/src/validate.ts
// Desired-state validation driven by ONCE's provider registry, the port of
// io.github.getcolors.k3s.validate.
//
// Green renders its keys as Clojure keywords, so every message here carries the
// same leading colon — the three colours must report identical errors for one
// colors.yml.

import { parName } from "red/cli";
import type { Opts } from "red/workflow";
import { providers } from "package-once-red";

export { providers };

export const slots = ["provider-compute", "provider-dns", "provider-backend"];

export const supportedCompute = new Set(["hcloud"]);

interface ProviderEntry {
  required?: string[];
  secrets?: string[];
  tofuEnv?: Record<string, string>;
}

function entry(opts: Opts, slot: string): ProviderEntry | undefined {
  return (providers as Record<string, Record<string, ProviderEntry>>)[slot]?.[String(opts[slot])];
}

// Flat credential key to the environment variable consumed by OpenTofu.
export function tofuEnv(opts: Opts, slot: string): Record<string, string> {
  return entry(opts, slot)?.tofuEnv ?? {};
}

export function placeholder(value: unknown): boolean {
  return value == null ||
    (typeof value === "string" && (!value.trim() || value.toUpperCase() === "REPLACE_ME"));
}

function slotKeys(opts: Opts, field: "required" | "secrets"): string[] {
  return slots.flatMap((slot) => entry(opts, slot)?.[field] ?? []);
}

function missing(opts: Opts, keys: string[]): string[] {
  return keys.filter((key) => placeholder(opts[key]));
}

export const profilePar = parName("profile");

// Refuse the one environment overlay that could redirect remote state.
export function envErrors(env: Record<string, string | undefined>): string[] {
  return String(env[profilePar] ?? "").length
    ? [`${profilePar} is set. K3s takes its profile from colors.yml only — ` +
       "run from the project directory rather than overriding it."]
    : [];
}

const k3sVersionRe = /^v[0-9]+\.[0-9]+\.[0-9]+\+k3s[0-9]+$/;
const fluxVersionRe = /^v[0-9]+\.[0-9]+\.[0-9]+$/;
const httpsRepositoryRe = /^https:\/\/[A-Za-z0-9._~-]+(?:\/[A-Za-z0-9._~-]+)+(?:\.git)?$/;
const branchRe = /^[A-Za-z0-9._/-]+$/;
const pathRe = /^\.\/[A-Za-z0-9._/-]+$/;

// pr-str, for the unsupported-provider message: green prints the offending
// value through pr-str, which quotes strings and renders nil bare.
function prStr(value: unknown): string {
  if (value == null) return "nil";
  if (typeof value === "string") return JSON.stringify(value);
  return String(value);
}

// All credential-free validation errors.
export function stateErrors(opts: Opts): string[] {
  const compute = opts["provider-compute"];
  const errors: string[] = [];
  for (const key of missing(opts, ["profile", "workdir", "repository", "k3s-version",
                                   "flux-version", ...slotKeys(opts, "required")])) {
    errors.push(`:${key} is required`);
  }
  for (const slot of slots) {
    if (!((providers as Record<string, Record<string, unknown>>)[slot] ?? {})[String(opts[slot])]) {
      errors.push(`unsupported :${slot} ${prStr(opts[slot])}`);
    }
  }
  if ((providers as Record<string, Record<string, unknown>>)["provider-compute"][String(compute)] &&
      !supportedCompute.has(String(compute))) {
    errors.push(`unsupported :provider-compute ${prStr(compute)}` +
      " — K3s v1 supports hcloud only because it owns and tests that " +
      "provider's firewall");
  }
  if (typeof opts["compute-prevent-destroy"] !== "boolean") {
    errors.push(":compute-prevent-destroy must be true or false");
  }
  if (!placeholder(opts.repository) && !httpsRepositoryRe.test(String(opts.repository))) {
    errors.push(":repository must be a public HTTPS Git URL");
  }
  if (!placeholder(opts["k3s-version"]) && !k3sVersionRe.test(String(opts["k3s-version"]))) {
    errors.push(":k3s-version must look like v1.36.2+k3s1");
  }
  if (!placeholder(opts["flux-version"]) && !fluxVersionRe.test(String(opts["flux-version"]))) {
    errors.push(":flux-version must look like v2.9.2");
  }
  if (!(opts["repository-branch"] == null || branchRe.test(String(opts["repository-branch"])))) {
    errors.push(":repository-branch contains unsupported characters");
  }
  if (!(opts["repository-path"] == null || pathRe.test(String(opts["repository-path"])))) {
    errors.push(":repository-path must be a relative path beginning with ./");
  }
  return errors;
}

// Credentials required by the selected compute and backend providers.
export function secretErrors(opts: Opts): string[] {
  return [...new Set(missing(opts, slotKeys(opts, "secrets")))]
    .map((key) => `required credential is not set: ${parName(key)}`);
}

File: blue/src/package_k3s_blue/validate.py
"""Desired-state validation driven by ONCE's provider registry, the port of
io.github.getcolors.k3s.validate.

Green renders its keys as Clojure keywords, so every message here carries the
same leading colon — the three colours must report identical errors for one
colors.yml.
"""

from __future__ import annotations

import re

from blue.cli import par_name
from colors_compute import credential_requirements
from . import machine
providers={"provider-dns":{"no-infra":{"required":[],"secrets":[]},"cloudflare":{"required":[],"secrets":["cloudflare-api-token"]}}}

__all__ = ["providers"]

slots = ["provider-dns"]




def _entry(opts: dict, slot: str) -> dict | None:
    value = opts.get(slot)
    return providers.get(slot, {}).get(value) if isinstance(value, str) else None


def tofu_env(opts: dict, slot: str) -> dict[str, str]:
    """Flat credential key to the environment variable consumed by OpenTofu."""
    return (_entry(opts, slot) or {}).get("tofu-env", {})


def placeholder(x) -> bool:
    return x is None or (isinstance(x, str) and (not x.strip() or x.upper() == "REPLACE_ME"))


def _slot_keys(opts: dict, field: str) -> list[str]:
    return [key for slot in slots for key in (_entry(opts, slot) or {}).get(field, [])]


def _missing(opts: dict, keys: list[str]) -> list[str]:
    return [key for key in keys if placeholder(opts.get(key))]


profile_par = par_name("profile")


def env_errors(env: dict) -> list[str]:
    """Refuse the one environment overlay that could redirect remote state."""
    if str(env.get(profile_par) or ""):
        return [f"{profile_par} is set. K3s takes its profile from colors.yml only — "
                "run from the project directory rather than overriding it."]
    return []


_k3s_version_re = re.compile(r"v[0-9]+\.[0-9]+\.[0-9]+\+k3s[0-9]+")
_flux_version_re = re.compile(r"v[0-9]+\.[0-9]+\.[0-9]+")
_https_repository_re = re.compile(
    r"https://[A-Za-z0-9._~-]+(?:/[A-Za-z0-9._~-]+)+(?:\.git)?")
_branch_re = re.compile(r"[A-Za-z0-9._/-]+")
_path_re = re.compile(r"\./[A-Za-z0-9._/-]+")


def _pr_str(value) -> str:
    """pr-str, for the unsupported-provider message: green prints the offending
    value through pr-str, which quotes strings and renders nil bare."""
    if value is None:
        return "nil"
    if isinstance(value, str):
        return '"' + value.replace("\\", "\\\\").replace('"', '\\"') + '"'
    return str(value)


def state_errors(opts: dict) -> list[str]:
    """All credential-free validation errors."""
    compute = opts.get("provider-compute")
    errors: list[str] = machine.errors(opts)
    for key in _missing(opts, ["profile", "workdir", "repository", "k3s-version",
                               "flux-version", *_slot_keys(opts, "required")]):
        errors.append(f":{key} is required")
    for slot in slots:
        if opts.get(slot) not in providers.get(slot, {}):
            errors.append(f"unsupported :{slot} {_pr_str(opts.get(slot))}")
    if not isinstance(opts.get("compute-prevent-destroy"), bool):
        errors.append(":compute-prevent-destroy must be true or false")
    if (not placeholder(opts.get("repository"))
            and not _https_repository_re.fullmatch(str(opts.get("repository")))):
        errors.append(":repository must be a public HTTPS Git URL")
    if (not placeholder(opts.get("k3s-version"))
            and not _k3s_version_re.fullmatch(str(opts.get("k3s-version")))):
        errors.append(":k3s-version must look like v1.36.2+k3s1")
    if (not placeholder(opts.get("flux-version"))
            and not _flux_version_re.fullmatch(str(opts.get("flux-version")))):
        errors.append(":flux-version must look like v2.9.2")
    if not (opts.get("repository-branch") is None
            or _branch_re.fullmatch(str(opts.get("repository-branch")))):
        errors.append(":repository-branch contains unsupported characters")
    if not (opts.get("repository-path") is None
            or _path_re.fullmatch(str(opts.get("repository-path")))):
        errors.append(":repository-path must be a relative path beginning with ./")
    return errors


def secret_errors(opts: dict) -> list[str]:
    """Credentials required by the selected compute and backend providers."""
    return [f"required credential is not set: {par_name(key)}"
            for key in dict.fromkeys(_missing(opts, [*_slot_keys(opts, "secrets"),*[v.removeprefix("COLORS_PAR_").lower().replace("_","-") for v in credential_requirements(opts)]]))]
